### ISSUE TITLE
ensure all methods accept a callback as last argument and only in last

### ISSUE THREAD

**Issue by vvo** [algolia/algoliasearch-client-js#43]
## tl;dr;

Example:
`index.search` signature is cryptic and embeds strange parameters like delay. It does not makes it clear that we must have a callback in last parameter.

Other signatures does not always have callback as last parameter.

## TODO

- [x] remove delay parameter in .search(), should be implemented by the user if needed
    easily implemented with [lodash.debounce](https://lodash.com/docs#debounce) for example
- [x] search can be called with (query, params) or (params)
- [x] check all method signatures are using a callback as last parameter
- [x] rework addUserKeyWithValidity signature
